### PR TITLE
Feat/tflog callback

### PR DESCRIPTION
**Pull request by cxzhang4**
* [x] Update NEWS.md
* [x] Refactor logging functionality to avoid repeating code
* [x] Update style (period at the beginning of private functions)
* [x] Remove code that uses new R features
* [x] Refactor to use `map_lgl()` and avoid extraneous function calls
* [x] Remove magic numbers
* [x] Remove extraneous comments
* [x] Add documentation describing how to use TensorBoard
* [x] Log the current epoch as the step index
* [x] add a parameter that determines whether we log the individual training losses (sometimes only want validation)
* [x] Factor out the helper function for counting events

### CODE REVIEW

File: R/CallbackSetTB.R [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        map(names(self$ctx$measures_valid), private$log_valid_score)
+      }
+    }
+  ),
+  private = list(
+    # TODO: refactor into a single function with the following signature
+    # log_score = function(prefix, measure_name, score) {
+    #
+    # },
+    log_valid_score = function(measure_name) {
```
Comment (cxzhang4) [inline R/CallbackSetTB.R:48]: prepend a period for private methods

File: R/CallbackSetTB.R [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      valid_score = list(self$ctx$last_scores_valid[[measure_name]])
+      names(valid_score) = paste0("valid.", measure_name)
+      with_logdir(self$path, {
+        do.call(log_event, valid_score)
+      })
+    },
+    log_train_score = function() {
+      # TODO: remind ourselves why we wanted to display last_loss and not last_scores_train
```
Comment (cxzhang4) [inline R/CallbackSetTB.R:56]: add logging of the training measures (implementation similar to validation, refactor both to use a single function)

File: tests/testthat/test_CallbackSetTB.R [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+            eval_freq = eval_freq
+  )
+  mlp$param_set$set_values(cb.tb.path = pth0)
+
+  mlp$train(task)
+
+  events = mlr3misc::map(collect_events(pth0)$summary, unlist)
+
+  n_last_loss_events = sum(unlist(mlr3misc::map(events, \(x) x["tag"] == "train.loss")))
+  n_valid_acc_events = sum(unlist(mlr3misc::map(events, \(x) x["tag"] == "valid.classif.acc")))
```
Comment (cxzhang4) [inline tests/testthat/test_CallbackSetTB.R:70]: map_lgl (or similar) will remove the need to unlist

File: tests/testthat/test_CallbackSetTB.R [hunk 4]
```diff
@@ -0,0 +1,83 @@
+library(tfevents)
+
+test_that("basic", {
```
Comment (cxzhang4) [inline tests/testthat/test_CallbackSetTB.R:3]: autotest

File: tests/testthat/test_CallbackSetTB.R [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  cb = t_clbk("tb", path = tempfile())
+  expect_torch_callback(cb, check_man = FALSE)
```
Comment (cxzhang4) [inline tests/testthat/test_CallbackSetTB.R:5]: check_man -> TRUE

File: R/CallbackSetTB.R [hunk 6]
```diff
@@ -0,0 +1,75 @@
+#' @title TensorBoard Logging Callback
+#'
+#' @name mlr_callback_set.tb
+#'
+#' @description
+#' Logs training loss and validation measures as events that can be tracked using TensorBoard.
```
[the remark this answers is not in the document]
Comment (cxzhang4) [reply R/CallbackSetTB.R:6]: Mention that this requires the `tensorflow` package

File: R/CallbackSetTB.R [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    initialize = function(path) {
+      self$path = assert_path_for_output(path)
+      if (!dir.exists(path)) {
+        dir.create(path, recursive = TRUE)
+      }
+    },
+    #' @description
+    #' Logs the training loss and validation measures as TensorFlow events.
+    #' Meaningful changes happen at the end of each epoch.
+    #' Notably NOT on_batch_valid_end, since there are no gradient steps between validation batches,
```
[the remark this answers is not in the document]
Comment (cxzhang4) [reply R/CallbackSetTB.R:31]: Consider keeping it as a comment

File: R/CallbackSetTB.R [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    #' and therefore differences are due to randomness
+    # TODO: display the appropriate x axis with its label in TensorBoard
```
Comment (cxzhang4) [inline R/CallbackSetTB.R:33]: Use the step argument in log_event calls

File: R/CallbackSetTB.R [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+      })
+    }
+  )
+)
+
+
+mlr3torch_callbacks$add("tb", function() {
+  TorchCallback$new(
+    callback_generator = CallbackSetTB,
+    param_set = ps(
```
Comment (cxzhang4) [inline R/CallbackSetTB.R:68]: add a parameter that determines whether we log the individual training losses (sometimes only want validation)

**Review by cxzhang4 — COMMENTED** [review]
See individual comments

**Comment by cxzhang4** [comment]
The conversations left open are the ones that I see as slightly more ambiguous (we might want to discuss design, or look more closely to confirm that they have truly been addressed)

File: NEWS.md [hunk 10]
```diff
@@ -1,5 +1,6 @@
 # mlr3torch dev
 
+* Added `CallbackSetTB`, which allows logging that can be viewed by TensorBoard.
```
Comment (cxzhang4) [inline NEWS.md:3]: Should the most recent news be at the top?

File: R/CallbackSetTB.R (hunk 8 above)
Comment (sebffischer) [reply R/CallbackSetTB.R]: What's ambiguous here?

File: R/CallbackSetTB.R (hunk 9 above)
Comment (sebffischer) [reply R/CallbackSetTB.R:77]: What's ambiguous here

File: NEWS.md (hunk 10 above)
Comment (sebffischer) [reply NEWS.md:3]: Usually I add them at the bottokm

File: R/CallbackSetTB.R (hunk 9 above)
Comment (cxzhang4) [reply R/CallbackSetTB.R:77]: I guess nothing major, just if you were picky about the name

File: R/CallbackSetTB.R (hunk 8 above)
Comment (cxzhang4) [reply R/CallbackSetTB.R]: I only did some basic sanity checking (basically just looked at the test case in TensorBoard), so wasn't 100% confident in the correctness at the time of writing. But I think there's not a whole lot to mess up here

File: R/CallbackSetTB.R [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  )
+)
+
+#' @include TorchCallback.R
+mlr3torch_callbacks$add("tb", function() {
+  TorchCallback$new(
+    callback_generator = CallbackSetTB,
+    param_set = ps(
+      path           = p_uty(tags = c("train", "required")),
+      log_train_loss = p_lgl(tags = c("train", "required"))
```
Comment (sebffischer) [inline R/CallbackSetTB.R:79]:
```suggestion
      log_train_loss = p_lgl(tags = c("train", "required"), init = FALSE)
```

File: R/CallbackSetTB.R [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    ),
+    id = "tb",
```
Comment (sebffischer) [inline R/CallbackSetTB.R:81]:
```suggestion
    id = "tb",
    packages = "tfevents",
```

File: R/CallbackSetTB.R [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+  lock_objects = FALSE,
+  public = list(
+    #' @description
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    initialize = function(path, log_train_loss) {
+      self$path = assert_path_for_output(path)
+      if (!dir.exists(path)) {
+        dir.create(path, recursive = TRUE)
+      }
+      self$log_train_loss = assert_logical(log_train_loss)
```
Comment (sebffischer) [inline R/CallbackSetTB.R:30]:
```suggestion
      self$log_train_loss = assert_flag(log_train_loss)
```

Comment (sebffischer) [reply R/CallbackSetTB.R:30]: logical vectors don't need to have length 1 and can also be NA

File: R/CallbackSetTB.R [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+      }
+
+      if (length(self$ctx$last_scores_valid)) {
+        map(names(self$ctx$measures_valid), private$.log_valid_score)
+      }
+    }
+  ),
+  private = list(
+    .log_score = function(prefix, measure_name, score) {
+      event_list = list(score, self$ctx$epoch)
```
Comment (sebffischer) [inline R/CallbackSetTB.R:50]:
```suggestion
      event_list = set_names(list(score, self$ctx$epoch), c(paste0(prefix, measure_name), "step"))
```

File: R/CallbackSetTB.R [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+      names(event_list) = c(paste0(prefix, measure_name), "step")
```
Comment (sebffischer) [inline R/CallbackSetTB.R:51]:
```suggestion
```

File: tests/testthat/test_CallbackSetTB.R [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+  events = mlr3misc::map(collect_events(pth0)$summary, unlist)
+
+  n_train_loss_events = sum(unlist(mlr3misc::map(events, event_tag_is, tag_name = "train.loss")))
+  n_train_acc_events = sum(unlist(mlr3misc::map(events, event_tag_is, tag_name = "train.classif.acc")))
+  n_train_ce_events = sum(unlist(mlr3misc::map(events, event_tag_is, tag_name = "train.classif.ce")))
+  n_valid_acc_events = sum(unlist(mlr3misc::map(events, event_tag_is, tag_name = "valid.classif.acc")))
+  n_valid_ce_events = sum(unlist(mlr3misc::map(events, event_tag_is, tag_name = "valid.classif.ce")))
+
+  # TODO: refactor to expect a specific ordering of the events list, not just the right counts
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:46]: I guess this can be deleted?

File: R/CallbackSetTB.R [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    },
+    #' @description
+    #' Logs the training loss, training measures, and validation measures as TensorFlow events.
+    on_epoch_end = function() {
+      if (self$log_train_loss) {
+        private$.log_train_loss()
+      }
+
+      if (length(self$ctx$last_scores_train)) {
+        map(names(self$ctx$measures_train), private$.log_train_score)
```
Comment (sebffischer) [inline R/CallbackSetTB.R:40]:
```suggestion
        walk(names(self$ctx$measures_train), private$.log_train_score)
```

Comment (sebffischer) [reply R/CallbackSetTB.R:40]: `map` returns something but here we only call into the function for side-effects -> use `walk()`

Comment (sebffischer) [reply R/CallbackSetTB.R:40]: (which returns `NULL`)

File: R/CallbackSetTB.R [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+      }
+
+      if (length(self$ctx$last_scores_valid)) {
+        map(names(self$ctx$measures_valid), private$.log_valid_score)
```
Comment (sebffischer) [inline R/CallbackSetTB.R:44]:
```suggestion
        walk(names(self$ctx$measures_valid), private$.log_valid_score)
```

File: R/CallbackSetTB.R [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+    #' Creates a new instance of this [R6][R6::R6Class] class.
+    initialize = function(path, log_train_loss) {
+      self$path = assert_path_for_output(path)
+      if (!dir.exists(path)) {
+        dir.create(path, recursive = TRUE)
+      }
+      self$log_train_loss = assert_logical(log_train_loss)
+    },
+    #' @description
+    #' Logs the training loss, training measures, and validation measures as TensorFlow events.
```
Comment (sebffischer) [inline R/CallbackSetTB.R:33]:
```suggestion
    #' Logs the training loss, training measures, and validation measures as TensorBoard events.
```

File: tests/testthat/test_CallbackSetTB.R [hunk 20]
```diff
@@ -0,0 +1,131 @@
+library(tfevents)
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:1]: I don't this this should be here.

File: tests/testthat/test_CallbackSetTB.R [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+
+event_tag_is = function(event, tag_name) {
+  ifelse(is.null(event), FALSE, event["tag"] == tag_name)
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:4]:
```suggestion
  if (is.null(event)) FALSE else event["tag"] == tag_name
```

Comment (sebffischer) [reply tests/testthat/test_CallbackSetTB.R:4]: `ifelse` in R is vectorized

File: tests/testthat/test_CallbackSetTB.R [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+test_that("autotest", {
+  cb = t_clbk("tb", path = tempfile(), log_train_loss = TRUE)
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:8]:
```suggestion
  cb = t_clbk("tb", path = tempfile())
```

Comment (sebffischer) [reply tests/testthat/test_CallbackSetTB.R:8]: Now that we have a default no need to specify `log_train_loss`

File: tests/testthat/test_CallbackSetTB.R [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+  expect_torch_callback(cb, check_man = TRUE)
+})
+
+# TODO: investigate what's happening when there is only a single epoch (why don't we log anything?)
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:12]: Is this now done?

File: tests/testthat/test_CallbackSetTB.R [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+test_that("a simple example works", {
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:13]:
```suggestion
test_that("metrics are logged correctly", {
```

File: tests/testthat/test_CallbackSetTB.R [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+  cb = t_clbk("tb")
+
+  task = tsk("iris")
+  n_epochs = 10
+  batch_size = 50
+  neurons = 200
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:19]:
```suggestion
  neurons = 10
```

File: tests/testthat/test_CallbackSetTB.R [hunk 26]
```diff
@@ -0,0 +1,131 @@
...
+  expect_torch_callback(cb, check_man = TRUE)
+})
+
+# TODO: investigate what's happening when there is only a single epoch (why don't we log anything?)
+test_that("a simple example works", {
+  cb = t_clbk("tb")
+
+  task = tsk("iris")
+  n_epochs = 10
+  batch_size = 50
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:18]:
```suggestion
  batch_size = 150
```

Comment (sebffischer) [reply tests/testthat/test_CallbackSetTB.R:18]: make tests run as fast as posible

File: tests/testthat/test_CallbackSetTB.R [hunk 27]
```diff
@@ -0,0 +1,131 @@
...
+  log_train_loss = TRUE
+
+  mlp = lrn("classif.mlp",
+            callbacks = cb,
+            epochs = n_epochs, batch_size = batch_size, neurons = neurons,
+            validate = 0.2,
+            measures_valid = msrs(c("classif.acc", "classif.ce")),
+            measures_train = msrs(c("classif.acc", "classif.ce"))
+  )
+  mlp$param_set$set_values(cb.tb.path = pth0)
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:32]:
```suggestion
  mlp$param_set$set_values(cb.tb.path = tempfile())
```

File: tests/testthat/test_CallbackSetTB.R [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+  cb = t_clbk("tb")
+
+  task = tsk("iris")
+  n_epochs = 10
+  batch_size = 50
+  neurons = 200
+
+  pth0 = tempfile()
+
+  log_train_loss = TRUE
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:23]:
```suggestion
```

File: tests/testthat/test_CallbackSetTB.R [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+# TODO: investigate what's happening when there is only a single epoch (why don't we log anything?)
+test_that("a simple example works", {
+  cb = t_clbk("tb")
+
+  task = tsk("iris")
+  n_epochs = 10
+  batch_size = 50
+  neurons = 200
+
+  pth0 = tempfile()
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:21]:
```suggestion
```

File: tests/testthat/test_CallbackSetTB.R [hunk 30]
```diff
@@ -0,0 +1,131 @@
...
+  )
+  mlp$param_set$set_values(cb.tb.path = pth0)
+
+  mlp$param_set$set_values(cb.tb.log_train_loss = log_train_loss)
+
+  mlp$train(task)
+
+  events = mlr3misc::map(collect_events(pth0)$summary, unlist)
+
+  n_train_loss_events = sum(unlist(mlr3misc::map(events, event_tag_is, tag_name = "train.loss")))
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:40]: Does `map_dbl` not also work here?

File: tests/testthat/test_CallbackSetTB.R [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+  mlp = lrn("classif.mlp",
+            callbacks = cb,
+            epochs = n_epochs, batch_size = batch_size, neurons = neurons,
+            validate = 0.2,
+            measures_valid = msrs(c("classif.acc", "classif.ce")),
+            measures_train = msrs(c("classif.acc", "classif.ce"))
+  )
+  mlp$param_set$set_values(cb.tb.path = pth0)
+
+  mlp$param_set$set_values(cb.tb.log_train_loss = log_train_loss)
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:34]:
```suggestion
  mlp$param_set$set_values(cb.tb.log_train_loss = TRUE)
```

File: tests/testthat/test_CallbackSetTB.R (hunk 28 above)
Comment (sebffischer) [reply tests/testthat/test_CallbackSetTB.R:23]: you bind to a variable but it is only used once

File: tests/testthat/test_CallbackSetTB.R (hunk 29 above)
Comment (sebffischer) [reply tests/testthat/test_CallbackSetTB.R:21]: you bind to a variable but it is only used once

File: tests/testthat/test_CallbackSetTB.R [hunk 32]
```diff
@@ -0,0 +1,131 @@
...
+            eval_freq = eval_freq
+  )
+  mlp$param_set$set_values(cb.tb.path = pth0)
+  mlp$param_set$set_values(cb.tb.log_train_loss = log_train_loss)
+
+  mlp$train(task)
+
+  events = mlr3misc::map(collect_events(pth0)$summary, unlist)
+
+  n_train_loss_events = sum(unlist(mlr3misc::map(events, event_tag_is, tag_name = "train.loss")))
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:82]: again `map_dbl`

File: tests/testthat/test_CallbackSetTB.R [hunk 33]
```diff
@@ -0,0 +1,131 @@
...
+  expect_equal(n_train_ce_events, n_epochs)
+  expect_equal(n_valid_acc_events, n_epochs)
+  expect_equal(n_valid_ce_events, n_epochs)
+})
+
+test_that("eval_freq works", {
+  cb = t_clbk("tb")
+
+  task = tsk("iris")
+  n_epochs = 9
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:58]: also here don't bind to variables if it is used only once

File: tests/testthat/test_CallbackSetTB.R [hunk 34]
```diff
@@ -0,0 +1,131 @@
...
+  expect_equal(n_train_ce_events, ceiling(n_epochs / eval_freq))
+  expect_equal(n_valid_acc_events, ceiling(n_epochs / eval_freq))
+  expect_equal(n_valid_ce_events, ceiling(n_epochs / eval_freq))
+})
+
+test_that("the flag for tracking the train loss works", {
+  cb = t_clbk("tb")
+
+  task = tsk("iris")
+  n_epochs = 10
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:99]: used only once

File: tests/testthat/test_CallbackSetTB.R [hunk 35]
```diff
@@ -0,0 +1,131 @@
...
+
+  log_train_loss = FALSE
+
+  pth0 = tempfile()
+
+  mlp = lrn("classif.mlp",
+            callbacks = cb,
+            epochs = n_epochs, batch_size = batch_size, neurons = neurons,
+            validate = 0.2,
+            measures_valid = msrs(c("classif.acc", "classif.ce")),
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:111]: here we only test the train loss, no need to pass measures, also no need to set `validate = 0.2`

File: tests/testthat/test_CallbackSetTB.R [hunk 36]
```diff
@@ -0,0 +1,131 @@
...
+            measures_train = msrs(c("classif.acc", "classif.ce"))
+  )
+  mlp$param_set$set_values(cb.tb.path = pth0)
+  mlp$param_set$set_values(cb.tb.log_train_loss = log_train_loss)
+
+  mlp$train(task)
+
+  events = mlr3misc::map(collect_events(pth0)$summary, unlist)
+
+  n_train_loss_events = sum(unlist(mlr3misc::map(events, event_tag_is, tag_name = "train.loss")))
```
Comment (sebffischer) [inline tests/testthat/test_CallbackSetTB.R:121]: use `map_lgl`

**Review by sebffischer — CHANGES_REQUESTED** [review]
Looking good!

Mostly minor style issues that are not so important in the tests but it's good to generally keep them in mind.

File: DESCRIPTION [hunk 37]
```diff
@@ -25,7 +25,12 @@ Authors@R:
              family = "Pfisterer",
              role = "ctb",
              email = "pfistererf@googlemail.com",
-             comment = c(ORCID = "0000-0001-8867-762X")))
+             comment = c(ORCID = "0000-0001-8867-762X")),
+      person(given = "Carson",
```
Comment (sebffischer) [inline DESCRIPTION:29]:
```suggestion
      aksldjalksdjkalsda
```

File: tests/testthat/test_CallbackSetTB.R [hunk 38]
```diff
@@ -0,0 +1,103 @@
+event_tag_is = function(event, tag_name) {
+  if (is.null(event)) FALSE else event["tag"] == tag_name
+}
+
+test_that("autotest", {
+  cb = t_clbk("tb", path = tempfile())
+  expect_torch_callback(cb, check_man = TRUE)
+})
+
+# TODO: investigate what's happening when there is only a single epoch (why don't we log anything?)
```
Comment (cxzhang4) [inline tests/testthat/test_CallbackSetTB.R:10]:
```suggestion
```

**Pull request merged**